Fix `__slots__` issues (overlaps, broken inheritance)

ariebovenberg opened the pull request:
#### References to other Issues or PRs

fixes #22922 

#### Brief description of what is fixed or changed

- `__slots__` added to some base classes
- specific slot names removed that were already declared on base classes

#### Other comments

I discovered the slots issues with [slotscheck](https://github.com/ariebovenberg/slotscheck), a tool I maintain. Of course, If you like, I can add it to tox/CI as I've done for [instagram/LibCST](https://github.com/Instagram/LibCST/pull/615), [dry-python/returns](https://github.com/dry-python/returns/pull/1233), and [aio-libs/aiohttp](https://github.com/aio-libs/aiohttp/pull/6547).

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* other
  * Fixed several issues with `__slots__` (overlaps and omission in base classes)
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
Looks like the `__slots__` are being used for something in codegen. Not sure if that's easy to fix.

ariebovenberg commented:
@oscarbenjamin ooo that'd be easy for me to overlook 👀 . I'll see what breaks then and let you know 😅

ThePauliPrinciple commented:
> Looks like the `__slots__` are being used for something in codegen. Not sure if that's easy to fix.

The role of `__slots__` in codegen/ast.py is documented in the docstring, maybe that is helpful to fix the issue.

oscarbenjamin commented on sympy/geometry/entity.py:
```diff
@@ -69,6 +69,8 @@ class GeometryEntity(Basic, EvalfMixin):
 
     """
 
+    __slots__ = ()
```
The mypy tests are failing because mypy can't figure out the "type" of slots here. I think it needs something like this to make mypy happy:
```diff
diff --git a/sympy/geometry/entity.py b/sympy/geometry/entity.py
index 1bd1722..5d541c1 100644
--- a/sympy/geometry/entity.py
+++ b/sympy/geometry/entity.py
@@ -19,6 +19,7 @@
 R3 are currently the only ambient spaces implemented.
 
 """
+from typing import Tuple as tTuple
 
 from sympy.core.basic import Basic
 from sympy.core.containers import Tuple
@@ -69,7 +70,7 @@ class GeometryEntity(Basic, EvalfMixin):
 
     """
 
-    __slots__ = ()
+    __slots__: tTuple[()] = ()
 
     def __cmp__(self, other):
         """Comparison of two GeometryEntities."""
```

ariebovenberg replied:
That'd make `mypy` happy, although I'd probably go for `tTuple[str, ...]` to ensure slots may be added later.

ariebovenberg commented:
@oscarbenjamin @ThePauliPrinciple thanks for the feedback. The mypy/authors issues are easy to fix, so only `codegen` remains.

I'd like to suggest using a separate classvar (like `__codegen_fields__` or something) instead of `__slots__` to indicate the fields for codegen. Using `__slots__` for codegen will cause overlaps. The Python docs have this to say about overlaps (emphasis mine):

> If a class defines a slot also defined in a base class, the instance variable defined by the base class slot is inaccessible (except by retrieving its descriptor directly from the base class). *This renders the meaning of the program undefined*. In the future, *a check may be added to prevent this*.

This tells us:
1. It may cause undefined behavior
2. It may be disallowed in future Python versions

Agreed to move "codegen" slots to another attribute, so `__slots__` overlaps can be removed?

ThePauliPrinciple commented:
@bjodah do you think there will be unintended side effects with this change?

I don't think we should use a "custom" double underscore attribute for this, something like `_attributes` or `_fields` makes more sense to me.

We could also remove overlapping values from `__slots__` and then check the MRO for any `CodegenAST` superclass and collect all `__slots__` from those.

ariebovenberg commented:
@ThePauliPrinciple agree that a single `_` name would be best. First, I'll investigate your suggestion on generating codegen slots from MRO.

ariebovenberg committed:
Fixed issues with __slots__ (overlaps and omission in base classes)

Across several modules, two types of slot problems were detected.

1) Overlaps
   A class redefines slots already present in a superclass.
   This reduces the memory savings from slots, as well as
   potentially introduces unpredictable behavior.

2) Omission in base classes
   A class defines slots, but one of its superclasses does not.
   This reduces the memory savings from slots, as well as allows
   `__dict__` to be created and non-slot attributes to be set.

Most of these issues were straightforward to fix, except in the `codegen`
module, which makes use of slots to generate constructors. Here a change
to the constructor logic was needed in order to solve the slots issues.

ariebovenberg commented:
I've adjusted the commits according to the development guide. What I realize now is that the branch name should reference the issue number. I can't make this change without opening a new pull request. What do you prefer?

ThePauliPrinciple commented on sympy/codegen/ast.py:
```diff
@@ -161,7 +161,17 @@ def _mk_Tuple(args):
+
+def _fields(cls):
+    if hasattr(cls, "_fields"):
+        yield from cls._fields
```
Why do we need this?

ariebovenberg replied:
Hmmm, I did indeed forget to explain.

The reason this is needed (and we cannot only use `__slots__`) is that there is no way to 'automatically' order inherited slots in the correct way.

One sensible way seemed from in reverse MRO:

```python
class A:
    __slots__ = ('x', 'y')

class B:
    __slots__ = ('z', )

_fields(B)  # x, y, z
```

This seems to be the rule followed by most `codegen.ast` members

However, some classes (like `FunctionPrototype`) add inherited slots in a different way:

```python
class Node:
    __slots__ = ("attrs", )

class FunctionPrototype(Node):
    __slots__ = ('return_type', 'name', 'parameters') + Node.__slots__

_fields(FunctionPrototype)  # attrs, return_type, name, parameters
# SHOULD BE return_type, name, parameters, attrs
```

I briefly considered just adopting the new order, but that would of course be a breaking change.

ariebovenberg replied:
If you prefer, I could also make `_fields` the only way to define fields (i.e. never fallback to `__slots__`). That'd simplify introspection but it'd require setting `_fields` on a lot of classes.

ThePauliPrinciple replied:
If going by the mro is causing ordering issues then I think it is indeed better to simply define `_fields` fully. It would also make it clearer what to do for a future author who wants to create another codegen ast class.

oscarbenjamin commented:
> the branch name should reference the issue number

It's fine to use a different branch name. Using the issue number can be helpful so you remember what the branch is but ultimately it's just a branch in your fork rather than in the sympy repo so you can call it what you want.

asmeurer commented:
> > the branch name should reference the issue number
> 
> It's fine to use a different branch name. Using the issue number can be helpful so you remember what the branch is but ultimately it's just a branch in your fork rather than in the sympy repo so you can call it what you want.

Does it still say that in the contributing guide? Can we remove this? Typing out branch numbers is super annoying whenever I checkout someone's branch. I'd much prefer short to the point branch names.

ariebovenberg commented:
> > > the branch name should reference the issue number
> > 
> > 
> > It's fine to use a different branch name. Using the issue number can be helpful so you remember what the branch is but ultimately it's just a branch in your fork rather than in the sympy repo so you can call it what you want.
> 
> Does it still say that in the contributing guide? Can we remove this? Typing out branch numbers is super annoying whenever I checkout someone's branch. I'd much prefer short to the point branch names.

It's documented here, if you want to remove it: https://github.com/sympy/sympy/wiki/Development-workflow#branch-names

bjodah commented:
> @bjodah do you think there will be unintended side effects with this change?

No, I think it's fine as long as tests pass. I was not aware of the finer points of overriding `__slots__` in subclasses, nice to see there's a tool to teach us to do better :+1:.

ThePauliPrinciple commented:
I'd be in favour of merging this PR and have a seperate issue/PR for discussing the CI.

oscarbenjamin commented:
It's not really clear to me how useful `__slots__` actually is. I know it used to be an attempt to save memory but other ways of saving memory in CPython have been developed since then. As I understand it using `__slots__` is not good for performance in pypy.

Is there a strong advantage in using `__slots__` at all?

asmeurer commented:
I'm OK with adding this to the CI. 

> It's documented here, if you want to remove it: sympy/sympy/wiki/Development-workflow#branch-names

I removed it and replaced the example branch name with something better.

asmeurer commented:
> Is there a strong advantage in using __slots__ at all?

I think `__slots__` can also positively impact performance. It might be worth benchmarking, both the speed and memory impacts of having vs. not having `__slots__`.

As far as downsides, I'm not aware of anyone ever complaining about it. I've seen it cause weird issues with metaclasses, but I think that usually means you are doing something wrong anyway. It also makes SymPy objects unable to have weak references (without `__weakref__` in `__slots__`). And obviously it makes it impossible to assign random attributes to SymPy objects, but that's usually not something users should do anyway.

ariebovenberg commented:
> I think __slots__ can also positively impact performance. It might be worth benchmarking, both the speed and memory impacts of having vs. not having __slots__.

Indeed, the two advantages are:
1. better performance of attribute access
2. reduced memory footprint

There are some benchmarks floating around that should be easy to find. Of course, it's best to measure your specific case.

I'm not sure of the impact on other Python implementations like `pypy` though 🤔

oscargus commented:
I like the idea from @bjodah , merge this and discuss `slotcheck` in another issue.

But for what it is worth: independent of if slots are a good idea or not, I do not think that we would like to misuse them, so if that can be prevented (and it is not overly complicated or time-consuming to run it), then it is a yes from me.

ariebovenberg commented:
@oscarbenjamin @ThePauliPrinciple @oscargus @asmeurer I've created a ticket to discuss `slotscheck` here: #23030.

oscargus commented:
Any objections against merging this? Seems like an improvement to me!

ariebovenberg commented:
@oscargus @oscarbenjamin @ThePauliPrinciple can this be merged?

oscarbenjamin commented:
This seems to be the cause of CI failures on the master branch.

The last commit is 30th Jan. After that length of time the PR should be closed and reopened to rerun CI before merging.

asmeurer commented:
Sorry about that. I didn't realize the most recent commit was from so long ago. We can do a quick revert of this unless @ariebovenberg can provide a quick fix.

oscarbenjamin commented:
Presumably some other PR has changed something affected by this. My guess is https://github.com/sympy/sympy/pull/23196

ariebovenberg commented:
@oscarbenjamin if that's the only failure, the fix is easy. I'll submit a new PR.

ariebovenberg commented:
@asmeurer @oscarbenjamin the issue has been fixed in #23252

oscarbenjamin commented:
Great, thanks.
